components/ledger/pkg/ledger/lock/lock.go

```diff
@@ -0,0 +1,59 @@
+package lock
+
+import (
+	"context"
+	"sync"
+)
+
+type Unlock func(ctx context.Context)
+
+type Locker interface {
+	Lock(ctx context.Context, ledger string, accounts ...string) (Unlock, error)
+}
+type LockerFn func(ctx context.Context, ledger string, accounts ...string) (Unlock, error)
+
+func (fn LockerFn) Lock(ctx context.Context, ledger string, accounts ...string) (Unlock, error) {
+	return fn(ctx, ledger, accounts...)
+}
+
+var NoOpLocker = LockerFn(func(ctx context.Context, ledger string, accounts ...string) (Unlock, error) {
+	return func(ctx context.Context) {}, nil
+})
+
+type InMemory struct {
+	globalLock sync.RWMutex
+	locks      map[string]*sync.Mutex
+}
+
+func (d *InMemory) Lock(ctx context.Context, ledger string, accounts ...string) (Unlock, error) {
+	d.globalLock.RLock()
+	lock, ok := d.locks[ledger]
+	d.globalLock.RUnlock()
+	if ok {
+		goto ret
+	}
+
+	d.globalLock.Lock()
+	lock, ok = d.locks[ledger] // Double check, the lock can have been acquired by another go routing between RUnlock and Lock
+	if !ok {
+		lock = &sync.Mutex{}
+		d.locks[ledger] = lock
+	}
+	d.globalLock.Unlock()
+ret:
```

Comment:
I'm not a big fan of goto, is it better to put the previous lock instead of the goto ?

Reply:
The previous lock is a read lock.
But i an modify this to remove the goto if you don't like it.